### PR TITLE
Extract output format to a class structure and add CLI parameter

### PR DESCRIPTION
**Pull request by kuhnroyal**
* add CLI format parameter with currently only one option (default)
* implement default format as class
* refactor rendering and optimise relative path & fatal handling
* only sort lints if the output format requires this (future JSON format does not need this)

I intend to create another PR with the Dart JSON output as follow up, but splitting this in a separate PR for easier review.
No test changes, output stays the same for now.

### CODE REVIEW

File: packages/custom_lint/bin/custom_lint.dart [hunk 1]
```diff
@@ -16,6 +16,19 @@ Future<void> entrypoint([List<String> args = const []]) async {
       help: 'Treat warning level issues as fatal',
       defaultsTo: true,
     )
+    ..addOption(
```
Comment (rrousselGit) [inline packages/custom_lint/bin/custom_lint.dart:19]: This flag seems useless at the moment to me.

Unless we have at least 2 formats, we shouldn't have it

Comment (kuhnroyal) [reply packages/custom_lint/bin/custom_lint.dart:19]: Sure, I can remove this and re-add it in the next PR with the JSON format.

Comment (kuhnroyal) [reply packages/custom_lint/bin/custom_lint.dart]: Done

**Comment by kuhnroyal** [comment]
I added the JSON output format to this PR. With the class separation it is pretty clean and should be fine to review.

File: packages/custom_lint/test/cli_test.dart [hunk 2]
```diff
@@ -228,43 +328,6 @@ Analyzing...
     );
   });
 
-  test('CLI lists warnings from all plugins and set exit code', () async {
```
Comment (kuhnroyal) [inline packages/custom_lint/test/cli_test.dart:231]: This test is a duplicate

File: packages/custom_lint/CHANGELOG.md [hunk 3]
```diff
@@ -1,3 +1,6 @@
+## UNRELEASED
```
Comment (rrousselGit) [inline packages/custom_lint/CHANGELOG.md:1]:
```suggestion
## Unreleased patch

```

File: packages/custom_lint/lib/src/output/json_output_format.dart [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+///
+/// Code is an adaption of the original Dart SDK JSON format.
+/// See: https://github.com/dart-lang/sdk/blob/main/pkg/dartdev/lib/src/commands/analyze.dart
+class JsonOutputFormat implements OutputFormat {
+  @override
+  void render({
+    required Iterable<AnalysisError> errors,
+    required Logger log,
+  }) {
+    Map<String, dynamic> location(
```
Comment (rrousselGit) [inline packages/custom_lint/lib/src/output/json_output_format.dart:18]: Those utils should be extracted to top-level private functions

And don't use => but { return}

File: packages/custom_lint/lib/src/output/json_output_format.dart [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+    Map<String, dynamic> range(
+      Map<String, dynamic> start,
+      Map<String, dynamic> end,
+    ) =>
+        {
+          'start': start,
+          'end': end,
+        };
+
+    final diagnostics = <Map<String, dynamic>>[];
```
Comment (rrousselGit) [inline packages/custom_lint/lib/src/output/json_output_format.dart:47]: No dynamic, use Object?

File: packages/custom_lint/test/cli_process_test.dart [hunk 6]
```diff
@@ -147,38 +147,82 @@ No issues found!
...
+            workingDirectory: app.path,
+          );
+
+          expect(trimDependencyOverridesWarning(process.stderr), isEmpty);
+
+          if (format == 'json') {
+            final dir = Directory(app.path).resolveSymbolicLinksSync();
+            expect(
+                process.stdout,
+                'Analyzing...\n\n${jsonEncode({
```
Comment (rrousselGit) [inline packages/custom_lint/test/cli_process_test.dart:181]: Move that jsonEncode to a separate line

File: packages/custom_lint/test/cli_test.dart [hunk 7]
```diff
@@ -32,87 +33,184 @@ Pattern progressMessage({required bool supportsAnsiEscapes}) {
...
+                  allOf(
+                    matches(
+                      progressMessage(
+                        supportsAnsiEscapes: ansi,
+                      ),
                     ),
-                  ),
-                  endsWith('''
+                    format == 'json'
+                        ? endsWith('${jsonEncode({
```
Comment (rrousselGit) [inline packages/custom_lint/test/cli_test.dart:106]: Again, construct those JSON strings out of the expectation

That expectation is way too long

Comment (rrousselGit) [reply packages/custom_lint/test/cli_test.dart:106]: I'd suggest extracting those JSONs to a dedicated .json file

File: packages/custom_lint/test/cli_test.dart (hunk 2 above)
Comment (rrousselGit) [reply packages/custom_lint/test/cli_test.dart:231]: Do you have the source of the other one?

**Comment by rrousselGit** [comment]
Overall LGTM, with some minor API/readability issues

File: packages/custom_lint/test/cli_test.dart (hunk 2 above)
Comment (kuhnroyal) [reply packages/custom_lint/test/cli_test.dart:231]: The one with the JSON above https://github.com/invertase/dart_custom_lint/pull/198/files#diff-0ec183ab5d11bde0260ed0f0d4a24da1edcd2c8bd6d3fb88f94566298721358fL70

File: packages/custom_lint/test/cli_test.dart (hunk 7 above)
Comment (kuhnroyal) [reply packages/custom_lint/test/cli_test.dart:106]: There is some string interpolation involved, so a JSON file is not great but could do a separate part file?

File: packages/custom_lint/lib/src/output/json_output_format.dart (hunk 4 above)
Comment (kuhnroyal) [reply packages/custom_lint/lib/src/output/json_output_format.dart:18]: The whole logic is pretty much a copy from the Dart SDK. I didn't want to change too much here so it is easier to adapt to changes in the SDK in the future. I only adapted minor things to please the linter.

I agree with all your change requests but left this intentionally the way it is.
If you think it is still better to adapt, I will change it.

File: packages/custom_lint/test/cli_test.dart (hunk 7 above)
Comment (rrousselGit) [reply packages/custom_lint/test/cli_test.dart:106]: A function would be great. No need to use a part directive for that

Comment (kuhnroyal) [reply packages/custom_lint/test/cli_test.dart]: Done

File: packages/custom_lint/lib/custom_lint.dart [hunk 8]
```diff
@@ -41,6 +38,7 @@ Future<void> customLint({
   required Directory workingDirectory,
   bool fatalInfos = true,
   bool fatalWarnings = true,
+  String format = 'default',
```
[the remark this answers is not in the document]
Comment (kuhnroyal) [reply packages/custom_lint/lib/custom_lint.dart]: Done

File: packages/custom_lint/test/cli_process_test.dart (hunk 6 above)
Comment (kuhnroyal) [reply packages/custom_lint/test/cli_process_test.dart]: Done

File: packages/custom_lint/lib/src/output/json_output_format.dart (hunk 5 above)
Comment (kuhnroyal) [reply packages/custom_lint/lib/src/output/json_output_format.dart]: Done

File: packages/custom_lint/lib/src/output/json_output_format.dart [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+      String filePath,
+      Map<String, dynamic> range,
+    ) =>
+        {
+          'file': filePath,
+          'range': range,
+        };
+
+    Map<String, dynamic> position(
+      int? offset,
```
[the remark this answers is not in the document]
Comment (kuhnroyal) [reply packages/custom_lint/lib/src/output/json_output_format.dart]: Done

**Comment by kuhnroyal** [comment]
Should be good to go

**Comment by rrousselGit** [comment]
Great, thanks!

**Pull request merged**